server/etcdserver/api/etcdhttp/health_test.go

```diff
@@ -49,136 +52,271 @@ func (s *fakeHealthServer) Config() config.ServerConfig {
-			if err != nil {
-				t.Errorf("fail serve http request %s %v", tt.healthCheckURL, err)
-			}
-			if res == nil {
-				t.Errorf("got nil http response with http request %s", tt.healthCheckURL)
-				return
-			}
-			if res.StatusCode != tt.expectStatusCode {
-				t.Errorf("want statusCode %d but got %d", tt.expectStatusCode, res.StatusCode)
-			}
-			health, err := parseHealthOutput(res.Body)
-			if err != nil {
-				t.Errorf("fail parse health check output %v", err)
+func TestDataCorruptionCheck(t *testing.T) {
+	be, _ := betesting.NewDefaultTmpBackend(t)
+	defer betesting.Close(t, be)
+	tests := []healthTestCase{
+		{
+			name:             "Live if CORRUPT alarm is on",
+			alarms:           []*pb.AlarmMember{{MemberID: uint64(0), Alarm: pb.AlarmType_CORRUPT}},
+			healthCheckURL:   "/livez",
+			expectStatusCode: http.StatusOK,
+			notInResult:      []string{"data_corruption"},
+		},
+		{
+			name:             "Not ready if CORRUPT alarm is on",
+			alarms:           []*pb.AlarmMember{{MemberID: uint64(0), Alarm: pb.AlarmType_CORRUPT}},
+			healthCheckURL:   "/readyz",
+			expectStatusCode: http.StatusServiceUnavailable,
+			inResult:         []string{"[-]data_corruption failed: Alarm active: CORRUPT", "[+]serializable_read ok"},
+		},
+		{
+			name:             "ready if CORRUPT alarm is not on",
+			alarms:           []*pb.AlarmMember{{MemberID: uint64(0), Alarm: pb.AlarmType_NOSPACE}},
+			healthCheckURL:   "/readyz",
+			expectStatusCode: http.StatusOK,
+		},
+		{
+			name:             "ready if CORRUPT alarm is excluded",
+			alarms:           []*pb.AlarmMember{{MemberID: uint64(0), Alarm: pb.AlarmType_CORRUPT}, {MemberID: uint64(0), Alarm: pb.AlarmType_NOSPACE}},
+			healthCheckURL:   "/readyz?exclude=data_corruption",
+			expectStatusCode: http.StatusOK,
+			inResult:         []string{"[+]serializable_read ok"},
+		},
+		{
+			name:             "Not ready if CORRUPT alarm is on",
+			alarms:           []*pb.AlarmMember{{MemberID: uint64(0), Alarm: pb.AlarmType_CORRUPT}},
+			healthCheckURL:   "/readyz?exclude=non_exist",
+			expectStatusCode: http.StatusServiceUnavailable,
+			inResult:         []string{"[-]data_corruption failed: Alarm active: CORRUPT", "[+]serializable_read ok"},
+		},
+		{
+			name:             "/ready/data_corruption error if CORRUPT alarm is on",
+			alarms:           []*pb.AlarmMember{{MemberID: uint64(0), Alarm: pb.AlarmType_CORRUPT}, {MemberID: uint64(0), Alarm: pb.AlarmType_NOSPACE}},
+			healthCheckURL:   "/readyz/data_corruption",
+			expectStatusCode: http.StatusServiceUnavailable,
+		},
+		{
+			name:             "/ready/serializable_read ok if CORRUPT alarm is on",
+			alarms:           []*pb.AlarmMember{{MemberID: uint64(0), Alarm: pb.AlarmType_CORRUPT}, {MemberID: uint64(0), Alarm: pb.AlarmType_NOSPACE}},
+			healthCheckURL:   "/readyz/serializable_read",
+			expectStatusCode: http.StatusOK,
+		},
```

Comment:
This case is misleading. I expect that in real etcd corrupt alarm will cause serializable read to fail. Not sure why we test serializable read in the `TestDataCorruptionCheck` test.

Reply:
Moved subpath check to a separate test.